File: sync-chat-client-py/src/client/proxy.py
"""Channel: UN socket TCP. Mismo rol que el Channel del cliente de Go
(y que ATMProxy en el cajero): el constructor solo recuerda la dirección,
open() conecta, close() libera.

Aquí la conversación no es ping-pong (una petición, una respuesta) sino que
el servidor puede escribir en cualquier momento, así que hace falta un
hilo de lectura continuo. "No bloqueante" se resuelve con el patrón
clásico de hilos: un hilo dedicado hace el recv() bloqueante a nivel de SO
mientras el resto del programa sigue vivo, y publica cada mensaje ya
parseado en una queue.Queue (thread-safe por diseño: es justamente el
mecanismo de sincronización entre el hilo lector y quien consume).

En este protocolo cada conversación tiene SU PROPIO Channel: el canal
grupal es uno, y cada chat privado abre otro independiente (con su propio
JOIN{from,to}). Channel no sabe si es "el grupal" o "un privado con X":
eso lo decide quien lo usa.
"""

from __future__ import annotations

import json
import queue
import socket
import threading
from typing import Optional

from src.model.message import Message


class Channel:
    def __init__(self, address: str):
        host, port = address.split(":")
        self._host = host
        self._port = int(port)
        self._sock: Optional[socket.socket] = None
        self._write_lock = threading.Lock()  # send() puede llamarse desde el hilo de UI mientras _read_loop sigue leyendo
        self.incoming: "queue.Queue[Optional[Message]]" = queue.Queue()

    def open(self) -> None:
        self._sock = socket.create_connection((self._host, self._port))
        threading.Thread(target=self._read_loop, daemon=True).start()

    def _read_loop(self) -> None:
        """Única hilo que lee este socket. Termina solo (dejando un
        centinela None en incoming) cuando el servidor cierra la conexión.
        """
        buffer = b""
        try:
            while True:
                chunk = self._sock.recv(4096)
                if not chunk:
                    return
                buffer += chunk
                while b"\n" in buffer:
                    line, buffer = buffer.split(b"\n", 1)
                    if not line:
                        continue
                    try:
                        self.incoming.put(Message.from_json(line.decode("utf-8")))
                    except (json.JSONDecodeError, UnicodeDecodeError):
                        continue  # línea corrupta: se descarta, no se tumba la conexión
        except OSError:
            return
        finally:
            self.incoming.put(None)

    def send(self, message: Message) -> None:
        data = (message.to_json() + "\n").encode("utf-8")
        with self._write_lock:
            self._sock.sendall(data)

    def close(self) -> None:
        if self._sock is not None:
            try:
                self._sock.close()
            finally:
                self._sock = None

File: sync-chat-client-go/src/client/proxy.go
package client

import (
	"bufio"
	"encoding/json"
	"net"
	"sync"

	"chat-client/src/model"
)

// Channel es UN socket TCP: el mismo rol que ATMProxy en el cajero, pero
// aquí la conversación no es ping-pong (una petición, una respuesta) sino
// que el servidor puede escribir en cualquier momento. Por eso, además de
// Open/Send/Close (igual que ATMProxy), hace falta una goroutine de
// lectura continua que deje cada mensaje en Incoming().
//
// En este protocolo cada conversación tiene SU PROPIO Channel: el canal
// grupal es uno, y cada chat privado abre otro independiente (con su
// propio JOIN{from,to}). Channel no sabe si es "el grupal" o "un privado
// con X": eso lo decide quien lo usa.
type Channel struct {
	address string
	conn    net.Conn
	reader  *bufio.Reader
	writer  *bufio.Writer
	wmu     sync.Mutex // Send puede llamarse desde la goroutine de UI mientras readLoop sigue leyendo

	incoming chan model.Message
}

// NewChannel NO conecta: solo recuerda a quién hay que llamar (igual que NewATMProxy).
func NewChannel(address string) *Channel {
	return &Channel{address: address}
}

// Open abre el socket y arranca la goroutine de lectura.
// Todavía no manda JOIN: eso lo decide el llamador (grupal vs. privado).
func (ch *Channel) Open() error {
	conn, err := net.Dial("tcp", ch.address)
	if err != nil {
		return err
	}
	ch.conn = conn
	ch.reader = bufio.NewReader(conn)
	ch.writer = bufio.NewWriter(conn)
	ch.incoming = make(chan model.Message, 64)

	go ch.readLoop()
	return nil
}

// readLoop es la única goroutine que lee este socket. "No bloqueante" se
// resuelve como en el cajero-cliente: el ReadBytes bloqueante a nivel de
// SO corre aparte, y el resto del programa sigue vivo consumiendo
// Incoming(). Termina sola (cerrando el channel) cuando el servidor cierra
// la conexión.
func (ch *Channel) readLoop() {
	defer close(ch.incoming)
	for {
		line, err := ch.reader.ReadBytes('\n')
		if err != nil {
			return
		}
		var msg model.Message
		if err := json.Unmarshal(line, &msg); err != nil {
			continue // línea corrupta: se descarta, no se tumba la conexión
		}
		ch.incoming <- msg
	}
}

// Incoming expone el canal de solo-lectura de mensajes ya parseados.
func (ch *Channel) Incoming() <-chan model.Message {
	return ch.incoming
}

func (ch *Channel) Send(msg model.Message) error {
	data, err := json.Marshal(msg)
	if err != nil {
		return err
	}

	ch.wmu.Lock()
	defer ch.wmu.Unlock()
	if _, err := ch.writer.Write(data); err != nil {
		return err
	}
	if err := ch.writer.WriteByte('\n'); err != nil {
		return err
	}
	return ch.writer.Flush()
}

// Close libera el extremo del cliente (igual que ATMProxy.Close).
func (ch *Channel) Close() {
	if ch.conn != nil {
		ch.conn.Close()
		ch.conn = nil
	}
}
